Feedback 1: Plexus folks on invariance and DMI blogs (#2)
mokyen/okyensourcesoftware: invariance.md

## Defining Invariance

In [C.2](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently), the Guidelines define invariance as:

> An invariant is a logical condition for the members of an object that a constructor must establish for the public member functions to assume. After the invariant is established (typically by a constructor) every member function can be called for the object. An invariant can be stated informally (e.g., in a comment) or more formally using Expects.
>
> If all data members can vary independently of each other, no invariant is possible.

I've read this explanation over and over, but it still never quite made sense in those terms. Let's propose an alternative definition of invariance:

> [!Important]  
> An invariant is a logical condition for the members of an object that can be expected to be true from the time the constructor exits to the time the destructor is called. These logical conditions fall into three categories.
>
> * A relationship between at least two data members such that all members cannot vary independently.

> > * A relationship between a data member and a resource (aka RAII).

Reviewer: I find the mention of RAII here a bit confusing. It almost feels like it's saying the category is RAII, and that doesn't seem right.

Author: RAII is a category, but I misunderstood that the invariance is the guarantee that RAII enforces. I clarified this

> * A property of a data member that must always be true.

Let's break down each of these.

## Relationships Between Members

Here's a simple example.

Consider a simple electrical circuit that is characterized by Ohm's Law:

> volts = current * resistance

> Ohm's law dictates that if you vary any one of the parameters, at least one other must also change. If the voltage increases, either the current or resistance must also increase. If the current decreases, either the resistance will increase or the voltage will decrease. They cannot be changed independently. If we were to code up an object that held these parameters, the parameters couldn't be public members. We'd need to enforce the invariance dictated by Ohm's Law.

Reviewer: You bring up that the parameters shouldn't be public members here, which is true, but also it took me a bit to get that your unspoken point was that they shouldn't be in a struct. Because the topic is "invariance", and it was a take off from the class vs struct, I wasn't in the headspace to be reflecting on this as an angle.

Author: I added a section at the beginning that aims to clarify the goal in understanding invariance.

      : m_amperage(current), m_resistance(resistance) {
    m_volts = current * resistance;
  }
  SimpleCiruit(Voltage volts, Resistance resistance)
      : m_volts(volts), m_resistance(resistance) {
    m_amperage = volts / resistance;
  }
  void setResistanceFixedVoltage(Resistance resistance) {
    m_resistance = resistance;
    m_amperage = m_volts / m_resistance;
  }
  void setResistanceFixedCurrent(Resistance resistance) {
    m_resistance = resistance;
    m_volts = m_amperage * m_resistance;
  }
  void setVoltsFixedCurrent(Voltage volts) {
    m_volts = volts;
    m_resistance = m_volts / m_amperage;
  }
  // ...
private:
  Voltage m_volts;
  Current m_amperage;
  Resistance m_resistance;
};
```

In our code, all the parameters are not necessarily linked in this way, and it's not necessarily the constructor that needs to enforce these relationships. For instance, consider this class:

```cpp
class Light {
public:
  void setBrightness(uint8_t brightness) {
    if (m_isLightOn) {
      m_brightness = brightness;
    }
  }
  void flipLightStatus() { m_isLightOn = !m_isLightOn; }

private:
  bool m_isLightOn;
  uint8_t m_brightness;
};
```

> The variable `m_isLightOn` on can change completely independently from `m_brightness`, but the reverse is not true. Therefore, these two pieces of data still have an invariance, or, said another way, the relationship between these two is an invariant.

Reviewer: ```suggestion
The variable `m_isLightOn` on can change completely independently from `m_brightness`, but the reverse is not true. Therefore, these two pieces of data still have an invariance, or, said another way, the relationship between these two contains an invariant.
```

To my understanding of `invariant`, it is the thing that cannot change, so it feels wrong to say the relationship in this case _is_ the invariant. I think it's more accurate to say the relationship _contains_ an invariant. Or maybe there's another way to put it that's better, but I haven't thought of it yet.

Author: I removed this last clause and just stated the invariance itself.

> ## Relationship Between a Member and Resource

Reviewer: I think I just don't get this section. Are we saying that owning a resource equates to establishing an invariant? Is the invariant that the owned resource exists? This section seems like a decent description of RAII, but doesn't tie it to invariance for me.

Author: Overhauled this whole section based on my new view on the invariance

When a data member needs to have a relationship with a "resource," the [Cpp Core Guidelines' recommended programming idiom is RAII](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#p8-dont-leak-any-resources).

cppreference.com [defines Resource Allocation Is Initialization](https://en.cppreference.com/w/cpp/language/raii) as:

> Resource Acquisition Is Initialization, or RAII, is a C++ programming technique that binds the life cycle of a resource that must be acquired before use (allocated heap memory, thread of execution, open socket, open file, locked mutex, disk space, database connection—anything that exists in limited supply) to the lifetime of an object.

> RAII establishes a relationship between the data member and the resource. Placing a file handle as a public member would be dangerous, as changing it could create a memory leak. RAII creates an invariance between the member and the resource.

Reviewer: This is another instance of mentioning public member that I assume is a reflection on the class vs struct. I had a hard time with why this was being pointed out initially.

Author: I called out how this necessitates a class

On a side note, I really prefer the term CADRe, or Constructor Acquires-Destructor Releases, as a better name for remembering RAII. It is just easier to remember the meaning. However, it's hard to change a term that's been around longer than the Super Nintendo.

## Property of a Data Member that Must Be True

The final type of invariance is similar to the first but isn't a relationship between data members. Instead, there's something that needs to be enforced about a member. Maybe our integer needs to always be even. Perhaps a string can only be a certain length. One of the most common places this is seen is in nontrivial getters or setters for an object that performs some enforcement.

Consider this example:

```cpp
class EvenNumber {
public:
  EvenNumber(int i) { setNumber(i); }
  void setNumber(int i) {
    if ((i % 2) != 0) { 
      throw std::runtime_error("Invalid input");
    } else {
      m_i = i;
    }
  }

private:
  int m_i;
};
```

Here, the invariance is a condition between a member and the constant *2*.

> I admit that this third category could be considered broad and open-ended. Developers could choose to interpret this to justify just about any data needing to be in a class. Most of the time, the manipulation of a data member is going to be compared to some constant data or literal. Developers should scrutinize if a piece of data needs that rule enforced before choosing to use a class.

Reviewer: Reading this I had a hard time deciding if "Developers" is referring to writers or users of the class with the invariant. The last sentence makes me think that it's users of the class. But you also started the paragraph in a way that I figured, as a consumer of this article, likely considering _making_ classes, that I should be reading this as a writer of classes.

Author: I felt uncomfortable with this section and that this was just too broad and open-ended. This third type of invariance is the newest one for me, and I'm struggling to come up with more concrete guidance. I left a TODO

> I admit that this third category could be considered broad and open-ended. Developers could choose to interpret this to justify just about any data needing to be in a class. Most of the time, the manipulation of a data member is going to be compared to some constant data or literal. Developers should scrutinize if a piece of data needs that rule enforced before choosing to use a class.

Reviewer: I don't understand `Developers could choose to interpret this to justify just about any data needing to be in a class.`

Author: Removed this entirely

## 'Invariant' Doesn't Mean Immutability

> Keep in mind that invariance means a relationship, not that the data cannot change. The term in software for something that cannot change is immutability.

Reviewer: ```suggestion
Keep in mind that invariance describes a relationship, not that the data cannot change. The term in software for something that cannot change is immutability.
```

Author: I rephrased this and made it more explicit

## So Why Do Invariants Matter?

[DMI](dmi.md) uses invariance as one of the key heuristics when determining if a class, a struct, or a free function should be implemented based on [C.2](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently) from the Guidelines. In that blog, I discuss more about applying invariance in the Decision Making Isolation philosophy.

## Complete Cpp Core Guideline on Invariance

With these simple examples in mind, here is what the Guidelines say about when to use a class in [C.2](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently):

> > ### <a name="Rc-struct"></a>C.2: Use `class` if the class has an invariant; use > `struct` if the data members can vary independently

Reviewer: ```suggestion
> ### <a name="Rc-struct"></a>C.2: Use `class` if the class has an invariant; use `struct` if the data members can vary independently
```

>
> but:
>
> ```cpp
>     class Date {
>     public:
>         // validate that {yy, mm, dd} is a valid date and initialize
>         Date(int yy, Month mm, char dd);
>         // ...
>     private:
>         int y;
>         Month m;
>         char d;    // day
>     };
> ```
>
> #### Note
>
> If a class has any `private` data, a user cannot completely initialize an object without the use of a constructor.
> Hence, the class definer will provide a constructor and must specify its meaning.
> This effectively means the definer needs to define an invariant.
>
> **See also**:
>
> * [define a class with private data as `class`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-class)
> * [Prefer to place the interface first in a class](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rl-order)
> * [minimize exposure of members](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-private)
> * [Avoid `protected` data](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-protected)
>
> #### Enforcement
>
> Look for `struct`s with all data private and `class`es with public members.

## But what about when my invariants or data change?

> One of the concerns that can arise from choosing to put data into a struct instead of a class with getters and setters is that those methods provide an interface that insulates callers if the underlying data needs to change. I agree that this not only can but absolutely *will* happen at some point. However, I think a couple of things can serve as protections. First, [DMI](dmi.md) gives guidance that keeps the scope of data small, meaning that the scope of changes will also be small. Second, changing from setting a variable directly to calling a getter or setter is a minor refactor. Third, the placement of data is often just a best guess at first, so the location, type, etc. of data is likely to change over time anyway. The setter/getter interface is also fluid in the first implementation. Modern IDEs continue to make this easier, and the introduction of AI will only continue this trend.

Reviewer: This first sentence is confusing to me. It sounds like "one of the concerns [...] is that those methods provide an interface that insulates callers if the underlying data needs to change" but the choice was putting "data into a struct instead of a class". So it reads as the problem is the struct because the class has methods that insulate callers from change, and I don't get that. From there I don't know what point the rest of the paragraph is making until we get to your "second" point, which suggests you're advocating converting structs to classes.
I don't know that this answers the question in the heading, and I don't have in my head a thought of a situation where that might occur off hand.

Author: I can completely see where I wasn't clear in my wording. I rewrote this as a more explicit problem paragraph and left most of the solution the same.

# What is Invariance?

The term 'invariant' is one of the linchpins of Decision Making Isolation. It is used in the [Cpp Core Guidelines](https://isocpp.github.io/CppCoreGuidelines) as the reason to use a class versus struct. But what does it really mean?

## Struct vs Class

Before we discuss invariance, let's take a look at what else the [Cpp Core Guidelines](https://isocpp.github.io/CppCoreGuidelines) say about the difference between a class and struct.

* [C.8](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c8-use-class-rather-than-struct-if-any-member-is-non-public): Use `class` rather than `struct` if any member is non-public

> While `class` and `struct` are relatively equivalent in C++, the Guidelines set the convention that structs are intended to only contain public members. Because all public data members can be directly modified, I believe it can be inferred that structs also shouldn't have public member functions. In the [DMI blog](dmi.md), we'll go more into the preference for free functions. For now, we'll assume that structs only have public data members and no functions.

Reviewer: Might be nice to define the initialism DMI here at its first use.

Author: Defined per your suggestion

> An invariant is a logical condition for the members of an object that a constructor must establish for the public member functions to assume. After the invariant is established (typically by a constructor) every member function can be called for the object. An invariant can be stated informally (e.g., in a comment) or more formally using Expects.
>
> If all data members can vary independently of each other, no invariant is possible.

I've read this explanation over and over, but it still never quite made sense in those terms. Let's propose an alternative definition of invariance:

> [!Important]  
> An invariant is a logical condition for the members of an object that can be expected to be true from the time the constructor exits to the time the destructor is called. These logical conditions fall into three categories.
>
> * A relationship between at least two data members such that all members cannot vary independently.
> * A guarantee that any function with access to a resource may be called at any time, i.e. implementing RAII.
> * A property of a data member that must always be true.

> A key for me is remembering that an invariance is **a logical condition**.

Reviewer: I like some of the discussion on this thread: https://softwareengineering.stackexchange.com/questions/32727/what-are-invariants-how-can-they-be-used-and-have-you-ever-used-it-in-your-pro

A couple of the things that caught my attention in the discussion:

* [citation](https://softwareengineering.stackexchange.com/a/364286)

  > An invariant is any logical rule that must be obeyed throughout the execution of your program that can be communicated to a human, but not to your compiler.

  This speaks to me because it ties in that you are designing in some constraints that may be important enough to validate at specified points during execution.

  Thinking about this, you could expand this sentence to say something like "... an invariance is **a logical condition** required of the class' purpose".

* I also like the example in the same comment regarding the gear shift. Perhaps there's another example that's actually software specific, but for the gear shift example, thinking about a situation where you design a class to represent it that treats each gear independently would require lots of invariant checks as opposed to the design where the single gear shift knob selects one and can only select one gear which has the same invariance but with less verification required. Maybe this is really out of scope the more I think about it, but it's an interesting perspective to think through.

    m_amperage = volts / resistance;
  }
  void setResistanceFixedVoltage(Resistance resistance) {
    m_resistance = resistance;
    m_amperage = m_volts / m_resistance;
  }
  void setResistanceFixedCurrent(Resistance resistance) {
    m_resistance = resistance;
    m_volts = m_amperage * m_resistance;
  }
  void setVoltsFixedCurrent(Voltage volts) {
    m_volts = volts;
    m_resistance = m_volts / m_amperage;
  }
  // ...
private:
  Voltage m_volts;
  Current m_amperage;
  Resistance m_resistance;
};
```

In our code, all the parameters are not necessarily linked in this way, and it's not necessarily the constructor that needs to enforce these relationships. For instance, consider this class:

```cpp
class Light {
public:
  void setBrightness(uint8_t brightness) {
    if (m_isLightOn) {
      m_brightness = brightness;
    }
  }
  void flipLightStatus() { m_isLightOn = !m_isLightOn; }

private:
  bool m_isLightOn;
  uint8_t m_brightness;
};
```

The variable `m_isLightOn` on can change completely independently from `m_brightness`, but the reverse is not true. Therefore, these two pieces of data still have an invariance: the brightness will only be updated when the light is on.

> ## Guaranteeing Manipulation of a Resource, or RAII

Reviewer: The term "manipulation" doesn't feel right here. "Guaranteeing lifetime of an owned resource", maybe?

Author: Changed per your suggestion

> Resources such as files, sockets, or heap memory need to be managed so that the right steps are made to open/close, connect/close, or allocate/deallocate these assets. It can be dangerous to assume that a caller will properly implement the correct initialization/deinitialization, so the [Resource Allocation is Initialization, or RAII,](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#p8-dont-leak-any-resources) technique was devised to encapsulate the assets. When RAII is utilized, the wrapping entity guarantees that any function that can access that object can be called from the end of its constructor until its destructor is called.

Reviewer: ```suggestion
Resources such as files, sockets, or heap memory need to be managed so that the right steps are made to open/close, connect/close, or allocate/deallocate these assets. It can be dangerous to assume that a caller will properly implement the correct initialization/deinitialization, so the [Resource Allocation is Initialization, or RAII,](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#p8-dont-leak-any-resources) technique was devised to encapsulate the assets. When implementing RAII, the wrapping entity (object) guarantees that any function that can access the underlying resource can be called from the end of the object's constructor until the object's destructor is called.
```

> Resources such as files, sockets, or heap memory need to be managed so that the right steps are made to open/close, connect/close, or allocate/deallocate these assets. It can be dangerous to assume that a caller will properly implement the correct initialization/deinitialization, so the [Resource Allocation is Initialization, or RAII,](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#p8-dont-leak-any-resources) technique was devised to encapsulate the assets. When RAII is utilized, the wrapping entity guarantees that any function that can access that object can be called from the end of its constructor until its destructor is called.

Reviewer: Another thought might be that the invariant here is the accessibility/usability of the resource.

This guarantee is our second form of invariance.

cppreference.com [defines Resource Allocation Is Initialization](https://en.cppreference.com/w/cpp/language/raii) as:

> Resource Acquisition Is Initialization, or RAII, is a C++ programming technique that binds the life cycle of a resource that must be acquired before use (allocated heap memory, thread of execution, open socket, open file, locked mutex, disk space, database connection—anything that exists in limited supply) to the lifetime of an object.
>
> RAII guarantees that the resource is available to any function that may access the object (resource availability is a [class invariant](https://en.wikipedia.org/wiki/Class_invariant), eliminating redundant runtime tests). It also guarantees that all resources are released when the lifetime of their controlling object ends, in reverse order of acquisition.

Since the goal of RAII is to insulate the resource from direct manipulation, this necessitates that we use a class instead of a struct based on our previous discussion.

> *On a side note, I really prefer the term CADRe, or Constructor Acquires-Destructor Releases, as a better name for remembering RAII. It is just easier to remember the meaning. However, it's hard to change a term that's been around longer than the Super Nintendo.*

Reviewer: I'm sorry, I had to look this up. [Super Nintendo (SNES)](https://en.wikipedia.org/wiki/Super_Nintendo_Entertainment_System) was released in Japan in 1990 and US in 1991. It seems that RAII as a name for this was introduced in 1994, even though the technique existed since at least the 80s.

Author: Hmm. I might have to rephrase this. My reference was the link below, but I can have some fun with this.

https://www.stroustrup.com/bs_faq.html#:~:text=I%20held%20back%20the%20introduction%20of%20exceptions%20into%20C%2B%2B%20for%20half%20a%20year%20until%20I%20discovered%20RAII%20(1988).%20RAII%20is%20an%20integral%20and%20necessary%20part%20of%20the%20C%2B%2B%20exception%20mechanism.

> *On a side note, I really prefer the term CADRe, or Constructor Acquires-Destructor Releases, as a better name for remembering RAII. It is just easier to remember the meaning. However, it's hard to change a term that's been around longer than the Super Nintendo.*

Author: ```suggestion
*On a side note, I really prefer the term CADRe, or Constructor Acquires-Destructor Releases, as a better name for remembering RAII. It is just easier to remember the meaning. However, it's hard to change a term that might have been around [longer than the Super Nintendo](https://www.stroustrup.com/bs_faq.html#:~:text=I%20held%20back%20the%20introduction%20of%20exceptions%20into%20C%2B%2B%20for%20half%20a%20year%20until%20I%20discovered%20RAII%20(1988).%20RAII%20is%20an%20integral%20and%20necessary%20part%20of%20the%20C%2B%2B%20exception%20mechanism) (or at least [older than the Nintendo 64](https://www.cardinalpeak.com/blog/raii-introduction)) .*
```

## Property of a Data Member that Must Be True

The final type of invariance is similar to the first but isn't a relationship between data members. Instead, some condition of a data member exists that must be enforced unrelated to another data member. Maybe our integer needs to always be even. Perhaps a string can only be a certain length. One of the most common places this is seen is in nontrivial getters or setters for an object that performs some enforcement.

Consider this example:

```cpp
class EvenNumber {
public:
  EvenNumber(int i) { setNumber(i); }
  void setNumber(int i) {
    if ((i % 2) != 0) { 
      throw std::runtime_error("Invalid input");
    } else {
      m_i = i;
    }
  }

private:
  int m_i;
};
```

> Here, the invariance is a condition between a member and the constant *2*.The most common case of this type of invariance is comparisons to a constant or literal, like the previous example. However, a similar case occurs when the invariance is a mathematical algorithm that must be applied to a member. For instance, a class could exists that stores an angle in radians but must convert the value to be in the range -2π to 2π. This is more involved than simply comparing to a constant, but it's a similar logical assertion that necessitates using a class.

Reviewer: ```suggestion
Here, the invariance is a condition between a member and the constant *2*. The most common case of this type of invariance is comparisons to a constant or literal, like in this example. However, a similar case occurs when the invariance is a mathematical algorithm that must be applied to a member. For instance, a class could exist that stores an angle in radians but must convert the value to be in the range -2π to 2π. This is more involved than simply comparing to a constant, but it's a similar logical assertion that necessitates using a class.
```

## 'Invariant' Doesn't Mean Immutability

Remember that invariance means a logical condition that must always be true, not that the data cannot change. The term in software for something that cannot change is *immutability*.

## So Why Do Invariants Matter?

[DMI](dmi.md) uses invariance as one of the key heuristics when determining if a class, a struct, or a free function should be implemented based on [C.2](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently) from the Guidelines. In that blog, I discuss more about applying invariance in the Decision Making Isolation philosophy.

## Complete Cpp Core Guideline on Invariance

With these simple examples in mind, here is what the Guidelines say about when to use a class in [C.2](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently):

> > ### <a name="Rc-struct"></a>C.2: Use `class` if the class has an invariant; use `struct` if the data members can vary independently

Reviewer: ```suggestion
> ### C.2: Use `class` if the class has an invariant; use `struct` if the data members can vary independently
```

Maybe doesn't matter, but this could be removed.

>
> but:
>
> ```cpp
>     class Date {
>     public:
>         // validate that {yy, mm, dd} is a valid date and initialize
>         Date(int yy, Month mm, char dd);
>         // ...
>     private:
>         int y;
>         Month m;
>         char d;    // day
>     };
> ```
>
> #### Note
>
> If a class has any `private` data, a user cannot completely initialize an object without the use of a constructor.
> Hence, the class definer will provide a constructor and must specify its meaning.
> This effectively means the definer needs to define an invariant.
>
> **See also**:
>
> * [define a class with private data as `class`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-class)
> * [Prefer to place the interface first in a class](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rl-order)
> * [minimize exposure of members](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-private)
> * [Avoid `protected` data](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-protected)
>
> #### Enforcement
>
> Look for `struct`s with all data private and `class`es with public members.

## But what about when my invariants or data change?

> One potential critique of these guidelines for choosing a struct versus a class is that public member functions provide an interface that isolates the private data members from the users of that object. If I have a class with a getter, I can change the type of that private data without modifying calls to the getter as long as the return type remains the same. If a struct is used and the data members are manipulated directly, any change to the types of those data members impacts the struct's consumer. Additionally, if a struct is used and we need to introduce an invariance for the private data (which would necessitate a switch from a struct to a class), then we'll also have to modify all the callers. Using a class is less likely to cause the callers of its member functions to change.

Reviewer: ```suggestion
One potential critique of these guidelines for choosing a struct versus a class is that public member functions provide an interface that isolates the private data members from the users of that object. If I have a class with a getter, I can change the type of that private data without modifying calls to the getter as long as the return type remains the same. If a struct is used and the data members are manipulated directly, any change to the types of those data members impacts the struct's users. Additionally, if a struct is used and we need to introduce an invariance (which would necessitate private data and therefore a switch from a struct to a class), then we'll also have to modify all the users. Using a class is less likely to cause the callers of its member functions to change.
```

> I agree that this is definitely a tradeoff. I think a couple of things can serve to minimize the impact of these scenarios. First, [DMI](dmi.md) gives guidance that keeps the scope of data small, meaning that the scope of updates when objects change will also be small. Second, changing from setting a variable directly to calling a getter or setter when a new invariance is introduced is a minor refactor. Third, the placement of data is often just a best guess at first, so the location, type, etc. of data is likely to change over time anyway. The setter/getter interface is also fluid early in development. Modern IDEs significantly simplify these types of refactors, and the introduction of AI will only continue this trend.

Reviewer: Not being familiar with DMI, I don't know whether it covers classes or structs, so I don't know in which context to frame the guidance. The "First, DMI..." sentence leads me to assume it's regarding structs. The "Second, ..." sentence throws me off because now we're talking about refactoring and I don't even know for sure from what to what yet. The "Third, ..." sentence contributes very little to my understanding; I can think about it and infer that refactoring is part of coding life. "The setter/getter" sentence feels like an after though that doesn't connect to the last sentence's "these types of refactors" maybe only because referring to the "setter/getter interface" as "fluid" doesn't clarify to me what you mean entirely.
